src/lib/io/sol-spi-impl-contiki-qmsi.c

```diff
@@ -0,0 +1,317 @@
+    qm_spi_config_t config;
+    struct {
+        qm_gpio_t port;
+        uint8_t pin;
+    } slave_select;
+    struct {
+        qm_spi_async_transfer_t xfer;
+        void (*cb)(void *data, struct sol_spi *spi, const uint8_t *tx,
+            uint8_t *rx, ssize_t status);
+        const void *data;
+        ssize_t status;
+    } xfer;
+};
+
+static process_event_t spi_irq_event;
+extern struct process soletta_app_process;
+
+static struct sol_spi *in_transfer[QM_SPI_NUM];
+
+static void
+spi_cb_dispatch(void *user_data, process_event_t ev, process_data_t ev_data)
+{
+    struct sol_spi *spi = (struct sol_spi *)ev_data;
+
+    qm_gpio_set_pin(spi->slave_select.port, spi->slave_select.pin);
+    spi->xfer.cb((void *)spi->xfer.data, spi, spi->xfer.xfer.tx,
+        spi->xfer.xfer.rx, spi->xfer.status);
+}
+
+static void
+spi_irq_dispatch(uint32_t id, ssize_t status)
+{
+    struct sol_spi *spi;
+
+    if ((spi = in_transfer[id]) == NULL)
+        return;
+    spi->xfer.status = status;
+    process_post(&soletta_app_process, spi_irq_event, spi);
+}
+
+static void
+tx_callback(uint32_t id, uint32_t len)
+{
+}
+
+static void
+rx_callback(uint32_t id, uint32_t len)
+{
+    spi_irq_dispatch(id, len);
+}
+
+static void
+err_callback(uint32_t id, qm_rc_t status)
+{
+    spi_irq_dispatch(id, -status);
+}
+
+static int
+spi_set_gpio_ss(struct sol_spi *spi)
+{
+    qm_gpio_port_config_t cfg;
+    uint32_t mask;
+    qm_rc_t ret;
+
+#if QUARK_SE
+    spi->slave_select.port = QM_GPIO_0;
+    switch (spi->bus) {
+    case QM_SPI_MST_0:
+        switch (spi->slave) {
+        case QM_SPI_SS_0:
+            spi->slave_select.pin = 24;
+            break;
+        case QM_SPI_SS_1:
+            spi->slave_select.pin = 25;
+            break;
+        case QM_SPI_SS_2:
+            spi->slave_select.pin = 26;
+            break;
+        case QM_SPI_SS_3:
+            spi->slave_select.pin = 27;
+            break;
+        default:
+            return -EINVAL;
+        }
+        break;
+    case QM_SPI_MST_1:
+        switch (spi->slave) {
+        case QM_SPI_SS_0:
+            spi->slave_select.pin = 11;
+            break;
+        case QM_SPI_SS_1:
+            spi->slave_select.pin = 12;
+            break;
+        case QM_SPI_SS_2:
+            spi->slave_select.pin = 13;
+            break;
+        case QM_SPI_SS_3:
+            spi->slave_select.pin = 14;
+            break;
+        default:
+            return -EINVAL;
+        }
+        break;
+    default:
+        return -EINVAL;
+    }
+#elif QUARK_D2000
+    spi->slave_select.port = QM_GPIO_0;
+    switch (spi->slave) {
+    case QM_SPI_SS_0:
+        spi->slave_select.pin = 0;
+        break;
+    case QM_SPI_SS_1:
+        spi->slave_select.pin = 1;
+        break;
+    case QM_SPI_SS_2:
+        spi->slave_select.pin = 2;
+        break;
+    case QM_SPI_SS_3:
+        spi->slave_select.pin = 3;
+        break;
+    default:
+        return -EINVAL;
+    }
+#endif
```

Comment:
I don't think so, not with the way pin-mux works and I don't feel comfortable adding more functions to it just for this, when, especially in the D2000 case, we want to add even less code.